tests/components/modbus/test_modbus_sensor.py

```diff
@@ -0,0 +1,337 @@
+"""The tests for the Modbus sensor component."""
+import unittest
+from unittest import mock
+
+from homeassistant.const import CONF_NAME, CONF_OFFSET, CONF_PLATFORM
+from homeassistant.components.modbus import DEFAULT_HUB, DOMAIN as MODBUS_DOMAIN
+from homeassistant.components.modbus.sensor import (
+    CONF_COUNT,
+    CONF_DATA_TYPE,
+    CONF_PRECISION,
+    CONF_REGISTER,
+    CONF_REGISTER_TYPE,
+    CONF_REGISTERS,
+    CONF_REVERSE_ORDER,
+    CONF_SCALE,
+    DATA_TYPE_CUSTOM,
+    DATA_TYPE_FLOAT,
+    DATA_TYPE_INT,
+    DATA_TYPE_UINT,
+    REGISTER_TYPE_HOLDING,
+    REGISTER_TYPE_INPUT,
+    PLATFORM_SCHEMA,
+    setup_platform,
+)
+from tests.common import get_test_home_assistant
+
+
+class TestModbusRegisterSensor(unittest.TestCase):
+    """Test the Modbus register sensor."""
+
+    def setUp(self):
+        """Initialize values for this testcase class."""
+        self.hass = get_test_home_assistant()
+        self.mock_hub = mock.MagicMock()
+        self.mock_hub.name = "hub"
+        self.hass.data[MODBUS_DOMAIN] = {DEFAULT_HUB: self.mock_hub}
+
+        self.common_register_config = {CONF_NAME: "test-config", CONF_REGISTER: 1234}
+
+    def tearDown(self):
+        """Stop everything that was started."""
+        self.hass.stop()
+
+    def run_test(self, register_config, register_words, expected):
+        """Helper for running tests for given config."""
+        config = PLATFORM_SCHEMA(
+            {CONF_PLATFORM: "modbus", CONF_REGISTERS: [register_config]}
+        )
+
+        class ReadResult:
+            def __init__(self, register_words):
+                self.registers = register_words
+
+        read_result = ReadResult(register_words)
+        register_type = config.get(CONF_REGISTERS)[0].get(CONF_REGISTER_TYPE)
+        if register_type == REGISTER_TYPE_INPUT:
+            self.mock_hub.read_input_registers.return_value = read_result
+        else:
+            self.mock_hub.read_holding_registers.return_value = read_result
+
+        sensors = []
+
+        def add_entities(entities):
+            nonlocal sensors
+            sensors = entities
+
+        setup_platform(self.hass, config, add_entities)
```

Comment:
Don't set up the platform directly. Use `async_setup_component` to set up the corresponding component which should set up the platform which should add an entity. Then assert the state of the entity by getting it from the core state machine.

The state is the truth of the entity. By writing the tests like this, they become more robust and we can change the entity implementation without changing the tests as long as the state remains the same.

Reply:
Ok. I had to mock also the integration part, but still the code is easier to read for those that do not know modbus sensor implementation details. I like it.